cmd/find/sizePrimary.go

```diff
@@ -0,0 +1,112 @@
+package cmdfind
+
+import (
+	"fmt"
+	"math"
+	"regexp"
+	"strconv"
+	"strings"
+
+	apitypes "github.com/puppetlabs/wash/api/types"
+)
+
+// Use bytesOf instead of a hash so that we generate a more readable
+// panic message
+func bytesOf(unit byte) uint64 {
+	switch unit {
+	case 'c':
+		return 1
+	case 'k':
+		return 1024
+	case 'M':
+		return 1024 * 1024
+	case 'G':
+		return 1024 * 1024 * 1024
+	case 'T':
+		return 1024 * 1024 * 1024 * 1024
+	case 'P':
+		return 1024 * 1024 * 1024 * 1024 * 1024
+	default:
+		panic(fmt.Sprintf("cmdfind.bytesOf received an unexpected unit %v", unit))
+	}
+}
+
+var sizeValueRegex = regexp.MustCompile(`^(\+|-)?((\d+)|(\d+[ckMGTP]))$`)
+
+// sizePrimary => -size (+|-)?((\d+)|(\d+[ckMGTP]))
+//
+// where c => character (byte), k => kilobyte, M => megabyte, G => gigabyte, T => terabyte, P => petabyte
+//
+// Example inputs:
+//   -size 2   (true if the entry's size in 512-byte blocks, rounded up, is 2)
+//   -size +2  (true if the entry's size in 512-byte blocks, rounded up, is greater than 2)
+//   -size -2  (true if the entry's size in 512-byte blocks, rounded up, is less than 2)
+//   -size +1k (true if the entry's size is greater than 1 kilobyte (1024 bytes))
+//
+//nolint
+var sizePrimary = newAtom([]string{"-size"}, func(tokens []string) (Predicate, []string, error) {
+	tokens = tokens[1:]
+	if len(tokens) == 0 {
+		return nil, nil, fmt.Errorf("-size: requires additional arguments")
+	}
+	v := tokens[0]
+	if !sizeValueRegex.MatchString(v) {
+		return nil, nil, fmt.Errorf("-size: %v: illegal size value", v)
+	}
+
+	cmp := v[0]
+	if cmp == '+' || cmp == '-' {
+		v = strings.TrimPrefix(v, string(cmp))
+	} else {
+		cmp = '='
+	}
+
+	p := func(e *apitypes.ListEntry) bool {
+		if !e.Attributes.HasSize() {
+			// TODO: If the user enters an expression like "-size 5 -o -true", then
+			// `wash find` will still include 'e' in its output even if 'e' does not have
+			// a size attribute. This leads to a weird UX in cases like `wash find /aws -size 5 -o true`
+			// b/c the output will contain things that don't have a `size` attribute
+			// (like EC2 instances). When we add the `state` attribute, then an expression
+			// like `wash find /aws -state running -o true` will include entries that don't
+			// have state, like S3 buckets and S3 objects.
+			//
+			// The more general issue here is how we handle entries that do not have the attributes
+			// required by the user's specified primaries when those primaries are or'ed
+			// with other primaries that could potentially return true for that entry. Do
+			// we ignore them? Do we still apply the predicate on them, even if that predicate
+			// may return true? We should figure this out soon.
```

Comment:
I'm thinking it'd be a better UX if `wash find`'s predicate always returned false for entries that do not have the attribute(s) specified by the given expression (so something like `wash find /aws -size +1 -o true` will never display EC2 instances because size doesn't make sense for them). 

We don't have to enforce this for _all_ attributes, e.g. I think entries without `ctime`, `mtime` or `atime` attributes should still be processed. Hmm, maybe we shouldn't enforce it for `ctime`, but we should for the other attributes?

If the TODO here is not an issue or is something we should address if users become annoyed by it, I'm happy to remove it too.

Reply:
Why would you use `-o true`? I'm not clear on that.

Reply:
Oh, the real use-case could be something llike `find /aws -size +1 -o mtime +1h` where an entry without a size attribute but w/ an mtime attribute could satisfy the predicate

Reply:
I'm comfortable saying that predicate returns false, but not going any further. We can do something more complicated if specific examples come up.

Reply:
Sounds good, I'll remove the TODO.